src/content-tags-drawer/ContentTagsCollapsible.jsx

```diff
@@ -36,20 +106,179 @@ import ContentTagsTree from './ContentTagsTree';
  * @param {string} taxonomyAndTagsData.contentTags.value - Value of applied Tag
  * @param {string} taxonomyAndTagsData.contentTags.lineage - Array of Tag's ancestors sorted (ancestor -> tag)
  */
-const ContentTagsCollapsible = ({ taxonomyAndTagsData }) => {
+const ContentTagsCollapsible = ({ contentId, taxonomyAndTagsData }) => {
   const intl = useIntl();
   const {
     id, name, contentTags,
   } = taxonomyAndTagsData;
 
+  // State to determine whether to update the backend or not
+  const [updatingTags, setUpdatingTags] = React.useState(false);
+  const mutation = useContentTaxonomyTagsMutation();
+
   const [isOpen, open, close] = useToggle(false);
   const [target, setTarget] = React.useState(null);
 
+  // Keeps track of the tree structure for the applied content tags passed
+  // in as a prop.
+  const [appliedContentTags, setAppliedContentTags] = React.useState({});
+
+  // Keeps track of the tree structure for tags that are add by selecting/unselecting
+  // tags in the dropdowns.
+  const [addedContentTags, setAddedContentTags] = React.useState({});
+
+  // To handle checking/unchecking tags in the SelectableBox
+  const [checkedTags, { add, remove }] = useCheckboxSetValues();
+
+  // Handles make requests to the backend whenever the checked tags change
+  React.useEffect(() => {
+    // We have this check because this hook is fired when the component first load
+    // and reloads (on refocus). We only want to make requests to the backend when
+    // the user is updating the tags.
+    if (updatingTags) {
+      setUpdatingTags(false);
+      const tags = checkedTags.map(t => decodeURIComponent(t.split(',').slice(-1)));
+      mutation.mutate({ contentId, taxonomyId: id, tags });
+    }
+  }, [contentId, id, checkedTags]);
+
+  const mergeTrees = (tree1, tree2) => {
+    const mergedTree = { ...tree1 };
+
+    const sortKeysAlphabetically = (obj) => {
+      const sortedObj = {};
+      Object.keys(obj)
+        .sort()
+        .forEach((key) => {
+          sortedObj[key] = obj[key];
+          if (obj[key] && typeof obj[key] === 'object') {
+            sortedObj[key].children = sortKeysAlphabetically(obj[key].children);
+          }
+        });
+      return sortedObj;
+    };
+
+    const mergeRecursively = (destination, source) => {
+      Object.entries(source).forEach(([key, sourceValue]) => {
+        const destinationValue = destination[key];
+
+        if (destinationValue && sourceValue && typeof destinationValue === 'object' && typeof sourceValue === 'object') {
+          mergeRecursively(destinationValue, sourceValue);
+        } else {
+          // eslint-disable-next-line no-param-reassign
+          destination[key] = sourceValue;
+        }
+      });
+    };
+
+    mergeRecursively(mergedTree, tree2);
+    return sortKeysAlphabetically(mergedTree);
+  };
+
+  // This converts the contentTags prop to the tree structure mentioned above
+  React.useEffect(() => {
```

Comment:
If this function is strictly manipulating data, it should be `useMemo`, not `useEffect`. Only use `useEffect` if you need to do something with the DOM elements.

Reply:
Got it, updated.